blog/2021-11-app-platform-updates.md

```diff
@@ -0,0 +1,416 @@
+
+```
+yarn d2-app-scripts start
+
+[WARNING] Found 2 versions of '@dhis2/app-runtime' in yarn.lock: 3.2.0, 3.2.4
+Package validation issues are ignored when running "d2-app-scripts start"
+HINT: Run "d2-app-scripts build" to automatically fix some of these issues
+```
+
+This shows that `@dhis2/app-runtime` has two different versions at the
+same time, which is not going to work.
+
+We have two different ways to resolve the problem.
+
+The first is that we can use the `build` command, which will prompt us
+if we would like to automatically fix the problem:
+
+```
+yarn d2-app-scripts build
+
+[WARNING] Found 2 versions of '@dhis2/app-runtime' in yarn.lock: 3.2.0, 3.2.4
+? There are duplicate dependencies in yarn.lock, would you like to correct them now? (Y/n)
+```
+
+Or, we can run the `deduplicate` command manually:
+
+```
+yarn d2-app-scripts deduplicate
+
+Run yarn install to deduplicate node_modules
+Done in 0.85s.
+```
+
+The `deduplicate` command fixes the lock-file, but you must run `yarn
+install` to re-install corrected dependencies to `node_modules`. The
+automatic resolution in `build` does the `yarn install` automatically.
+
+With the new features out of the way, time to turn an eye to the nuts
+and bolts of upgrading to the `8.x` series from a lower major version.
+
+There are some breaking changes that you will need to accomodate.
+
+### Breaking change: UI updated to `7.x`
+
+`@dhis2/ui` has been updated to `7.x` in the App Platform, so your
+application must also upgrade `@dhis2/ui` to `7.x`. We will go over the
+breaking changes in UI `7.x` [in a minute](#ui-7x).
+
+### Breaking change: styled-jsx updated to `4.x`
+
+`styled-jsx` has been updated to the `4.x` series throughout the entire
+App Platform, so your application will need to bump to `4.x` as well.
+This will not require any code changes in your application.
+
+### Breaking change: App Runtime updated to `3.x`
+
+`@dhis2/app-runtime` has been updated to `3.x` and you will need to do
+so on the application-side as well. We will go over the changes required
+for [App Runtime in the next section](#app-runtime-3x).
+
+### Breaking change: Jest updated to `27.x`
+
+The final change that may require changes in your application is that we
+have bumped Jest to `27.x`. Jest 27 has [changed various
+defaults](https://jestjs.io/blog/2021/05/25/jest-27) that may require
+flipping back in certain situations. On the App Platform side we bundle
+settings that work for a standard App Platform based application, but if
+you are doing special things in your tests, you may need to research if
+you need additional configuration.
+
+## App Runtime 3.x
+
+
+### New feature: `useDataQuery` caches and deduplicates queries
+
+The `useDataQuery` hook now automatically caches, and deduplicates,
+identical queries client-side.
+
+This means that multiple components requesting the same data within an
+arbitrary timeframe will not trigger multiple requests to the server.
```

Comment:
Technically the timeframe is not arbitrary, it deduplicates requests as long as a previous identical request is still in progress. I don't know if you kept it out for simplicity, but if you want to be specific that's what it does under the hood.

Reply:
Thanks, I left it out because I wasn't sure about the implementation details. I'll rephrase.

Reply:
Right, I was going off the proposal and generalised the mentioned "100ms" to "arbitrary" because I didn't know what the actual implementation ended up being.

Fixed in 1fac038.